rs/nns/governance/src/voting.rs

```diff
@@ -0,0 +1,432 @@
+};
+use ic_nns_common::pb::v1::{NeuronId, ProposalId};
+use std::{
+    cell::RefCell,
+    collections::{BTreeMap, BTreeSet, HashMap},
+};
+
+thread_local! {
+    static VOTING_STATE_MACHINES: RefCell<VotingStateMachines> = RefCell::new(VotingStateMachines::new());
+}
+
+struct VotingStateMachines {
+    // machines
+    machines: BTreeMap<ProposalId, ProposalVotingStateMachine>,
+}
+
+impl VotingStateMachines {
+    fn new() -> Self {
+        Self {
+            machines: BTreeMap::new(),
+        }
+    }
+
+    fn get_or_create_machine(
+        &mut self,
+        proposal_id: ProposalId,
+        topic: Topic,
+    ) -> &mut ProposalVotingStateMachine {
+        self.machines
+            .entry(proposal_id)
+            .or_insert_with(|| ProposalVotingStateMachine::try_new(proposal_id, topic).unwrap())
+    }
+
+    fn remove_if_done(&mut self, proposal_id: &ProposalId) {
+        if let Some(machine) = self.machines.get(proposal_id) {
+            if machine.is_done() {
+                self.machines.remove(proposal_id);
+            }
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Default)]
+struct ProposalVotingStateMachine {
+    // The proposal ID that is being voted on.
+    proposal_id: ProposalId,
+    // The topic of the proposal.
+    topic: Topic,
+    // Votes that have been cast before checking followees
+    neurons_to_check_followers: BTreeSet<NeuronId>,
+    // followers to process
+    followers_to_check: BTreeSet<NeuronId>,
+    // votes that need to be recorded in each neuron's recent_ballots
+    recent_neuron_ballots_to_record: BTreeMap<NeuronId, Vote>,
+}
+
+impl ProposalVotingStateMachine {
+    fn try_new(proposal_id: ProposalId, topic: Topic) -> Result<Self, String> {
+        if topic == Topic::Unspecified {
+            return Err("Topic must be specified".to_string());
+        }
+
+        Ok(Self {
+            proposal_id,
+            topic,
+            ..Default::default()
+        })
+    }
+
+    fn is_done(&self) -> bool {
+        self.neurons_to_check_followers.is_empty()
+            && self.followers_to_check.is_empty()
+            && self.recent_neuron_ballots_to_record.is_empty()
+    }
+
+    fn add_followers_to_check(
+        &mut self,
+        neuron_store: &NeuronStore,
+        voting_neuron: NeuronId,
+        topic: Topic,
+    ) {
+        self.followers_to_check
+            .extend(neuron_store.get_followers_by_followee_and_topic(voting_neuron, topic));
+        if ![Topic::Governance, Topic::SnsAndCommunityFund].contains(&topic) {
+            // Insert followers from 'Unspecified' (default followers)
+            self.followers_to_check.extend(
+                neuron_store.get_followers_by_followee_and_topic(voting_neuron, Topic::Unspecified),
+            );
+        }
+    }
+
+    fn cast_vote(&mut self, ballots: &mut HashMap<u64, Ballot>, neuron_id: NeuronId, vote: Vote) {
+        // It is not legitimate to cast an unspecified vote, but we cannot error out, so we
```

Comment:
update comment - we expecte Vote::Unspecified internally, but there's no action that we take, b/c this is to say "not voting"